rfd/00xx-pod-access-requests.md

```diff
@@ -0,0 +1,728 @@
+resources using pattern matching.
+The RBAC Role must include the resource name to grant access, but it requires
+knowing the Pod name in advance. In many cases, it is impossible to know them
+because Pods generated from Kubernetes objects like Deployments include random
+strings in their names and change each time the Pod is deleted or relocated from
+one node to the other.
+
+This RFD proposes an extension to Teleport RBAC where it will support
+managing access to Pods - including Pods names following a pattern - without
+relying on requesting access to Teleport Roles.
+
+## Role definition
+
+This RFD proposes an alternative to limit or extend user access to certain pods
+in a Kubernetes cluster from a Teleport Role.
+
+To achieve this, it proposes changes to Teleport's Roles to introduce a new
+Role version - `v6` - that extends the `v5` to include a field `kubernetes_pods`
+where it's possible to specify the Pod names to allow access.
+
+```yaml
+kind: role
+version: v6
+metadata:
+  name: my-kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_pods:
+    - name: A
+      namespace: namespace
+    - name: B
+      namespace: namespace
+    - name: C
+      namespace: other-namespace
+... other v5 fields
+```
+
+By default, `kubernetes_pods` is empty and will reject any access to Pods within
+the allowed clusters. To allow access to all Pods, the Role must include
+
+```yaml
+kubernetes_pods:
+- name: *
+  namespace: *
+```
+
+Restricting access to Pods will only apply to Kubernetes clusters that satisfy
+the `kubernetes_labels` values. This means that the following role:
+
+```yaml
+kind: role
+version: v6
+metadata:
+  name: restrict-prod-access-role
+spec:
+  allow:
+    kubernetes_labels:
+      "env": "prod"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_pods:
+    - name: deploy-*
+      namespace: *
+... other v5 fields
+```
+
+It will only restrict the Pod access to clusters with `env=prod`. If a user has another
+Role that grants him access to every Pod in any cluster - `kubernetes_labels: "*":"*"`
+and `kubernetes_pods: [name:*, namespace:*]`, he would still be able to access resources in
+other clusters, but for prod clusters, he could only access Pods prefixed by `deploy-`.
+
+## Details
+
+We investigated two different approaches to extend the Teleport's RBAC to control
+access to specific Pods. The ultimate goal of both approaches is to allow users
+to temporarily request access to certain Pods and for administrators to limit
+the scope of each Teleport Role without creating multiple Kubernetes RBAC Group
+per rule.
+
+The two approaches are considerably different, and we will detail each one to
+briefly describe the implementation details and their limitations.
+
+1. Teleport RBAC controls access to each Pod.
+
+When a user has at least one role that has a non-empty list of `kubernetes_pods`,
```

Comment:
Behavior when you have multiple roles granting access to different clusters (via kube_labels) is still not very clear. Do kube_resources matchers apply only to clusters matching kube_labels from the same role? Can you add a section with example when a user has 2 roles.